pkg/handler/handler.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	rc := regclient.New()
+	// iterate over all keys
+	for _, rKey := range providerRequest.Request.Keys {
+		// Providers should add a caching mechanism to avoid extra calls to external data sources.
+		// TODO: Determine time of cache length for policy decision
+
+		image, err := ref.New(rKey)
+		if err != nil {
+			utils.SendResponse(nil, fmt.Sprintf("unable to create ref for image %s: %v", rKey, err), w)
+			return
+		}
+
+		if image.Digest == "" {
+			klog.Warning("Image with tag was used, this is unsafe! ", "image", rKey)
+		}
+
+		// TODO: add support for multi-arch images
+		var digest string
+		manifest, err := rc.ManifestGet(req.Context(), image)
+		if err != nil {
+			utils.SendResponse(nil, fmt.Sprintf("unable to get manifest for image %s: %v", rKey, err), w)
+			continue
+		}
+		if manifest.IsList() {
+			klog.Info("Multi-platform image detected, looking up digest for current platform")
+			plat := platform.Platform{
+				Architecture: runtime.GOARCH,
+				OS:           runtime.GOOS,
+			}
+
+			desc, err := rcManifest.GetPlatformDesc(manifest, &plat)
+			if err != nil {
+				utils.SendResponse(nil, fmt.Sprintf("unable to get platform description for image %s: %v", rKey, err), w)
+				continue
+			}
+
+			digest = desc.Digest.String()
+		} else {
+			config, err := manifest.(rcManifest.Imager).GetConfig()
+			if err != nil {
+				utils.SendResponse(nil, fmt.Sprintf("unable to get config digest for image %s: %v", rKey, err), w)
+				continue
+			}
+			klog.Info("Using config digest")
+			digest = config.Digest.String()
+		}
+
+		digest = strings.TrimPrefix(digest, "sha256:")
+		klog.Info("Using resolved digest ", "digest", digest)
+
+		// Do verify
+		results := shareddata.UsePoliciesAndPublicKeys(func(policies map[string]dsse.Envelope, keys map[string]policy.PublicKey) (results *[]externaldata.Item) {
+			policyResults := make([]externaldata.Item, 0)
+
+			for policyName, policyEnv := range policies {
+				klog.Info("Verifying policy ", "name", policyName)
+
+				// Currently only support one signature per policy
+				policyKey := policyEnv.Signatures[0]
+
+				key, found := keys[policyKey.KeyID]
+				if !found {
+					klog.Error("unable to get policy key", "key", policyKey.KeyID)
+					policyResults = append(policyResults, externaldata.Item{
+						Key:   rKey,
+						Error: "Policy verification failed: no public key",
+					})
+					continue
+				}
+
+				// create a reader from the pem encoded public key
+				keyReader := bytes.NewReader(key.Key)
+				verifier, err := cryptoutil.NewVerifierFromReader(keyReader)
+				if err != nil {
+					klog.Error(err, "failed to create verifier")
+				}
+
+				subject := cryptoutil.DigestSet{cryptoutil.DigestValue{Hash: crypto.SHA256, GitOID: false}: digest}
+				subjects := []cryptoutil.DigestSet{subject}
+
+				var collectionSource source.Sourcer
+				memSource := source.NewMemorySource()
+
+				collectionSource = source.NewMultiSource(memSource, source.NewArchvistSource(vh.archivista))
```

Comment:
I don't think we need the MemorySource or MultiSource here. The intent behind the MultiSource is to allow users to provide attestations from disk with results from archivista for the policy engine. Since we're not dealing with attestations in memory we can skip out and just use the ArchivistSource (note to self: rename this in go-witness)

Reply:
thanks, I learned from this too

Reply:
Actually, this could be relevant for some work you're tracking Kris: You can supply multiple archivista sources to the multisource, so we should be able to search multiple archivista instances for matching attestations with this.